thread.c

```diff
@@ -5713,3 +5724,75 @@ rb_uninterruptible(VALUE (*b_proc)(VALUE), VALUE data)
     RUBY_VM_CHECK_INTS(cur_th->ec);
     return ret;
 }
+
+static rb_atomic_t rb_thread_storage_slots = 0;
+static rb_thread_storage_destructor *rb_thread_storage_destructors = NULL;
+
+rb_thread_storage_key_t
+rb_thread_storage_create_key(rb_thread_storage_destructor func)
+{
+    rb_thread_storage_key_t new_key = (rb_thread_storage_key_t)rb_thread_storage_slots;
+    rb_thread_storage_destructor *old_destructors = rb_thread_storage_destructors;
+    rb_thread_storage_destructor *destructors = ALLOC_N(rb_thread_storage_destructor, rb_thread_storage_slots + 1);
+    if (old_destructors) {
+        MEMCPY(destructors, old_destructors, rb_thread_storage_destructor, rb_thread_storage_slots);
+    }
+    destructors[new_key] = func;
+
+    ATOMIC_PTR_EXCHANGE(rb_thread_storage_destructors, destructors);
+    ATOMIC_INC(rb_thread_storage_slots);
+
+    if (old_destructors) {
+        // TODO: do we have an race condition here?
+        // e.g. is `rb_thread_storage_destructors[key]` atomic? I don't think so because it is `laod(rb_thread_storage_destructors + key)`
+        // The solution might be to not have key destructors and use thread EXIT hook instead.
```

Comment:
I'm quasi certain it's the case now, as I've seen a CI failure totally explained by it.

Another solution could be to have a `RB_THREAD_STORAGE_MAX` and to statically allocate that array (seems to be how `pthread_key_create` works. https://github.com/bminor/glibc/blob/b92a49359f33a461db080a33940d73f47c756126/nptl/pthread_keys.c#L22

Even a small number like `64` or `128` would be plenty. We may need to be able to return an error code though.